common/trainers/relevance_transfer_trainer.py

```diff
@@ -1,133 +1,154 @@
+                if self.config['n_gpu'] > 1:
+                    loss = loss.mean()
+                if self.config['gradient_accumulation_steps'] > 1:
+                    loss = loss / self.config['gradient_accumulation_steps']
+
+                loss.backward()
+
+                if (step + 1) % self.config['gradient_accumulation_steps'] == 0:
+                    self.optimizer.step()
+                    self.optimizer.zero_grad()
+                    self.iterations += 1
             else:
-                if 'resample' in self.config and self.config['resample']:
-                    indices = ImbalancedDatasetSampler(batch.text[0], batch.label).get_indices()
-                    batch_text = batch.text[0][indices]
-                    batch_lengths = batch.text[1][indices]
-                    batch_label = batch.label
-                else:
-                    batch_text = batch.text[0]
-                    batch_lengths = batch.text[1]
-                    batch_label = batch.label
+                # Clip gradients to address exploding gradients in LSTM
+                torch.nn.utils.clip_grad_norm_(self.model.parameters(), 25.0)
 
-            if hasattr(self.model, 'tar') and self.model.tar:
+                # Randomly sample equal number of positive and negative documents
+                self.train_loader.init_epoch()
                 if 'ignore_lengths' in self.config and self.config['ignore_lengths']:
-                    scores, rnn_outs = self.model(batch_text)
+                    if 'resample' in self.config and self.config['resample']:
+                        indices = ImbalancedDatasetSampler(batch.text, batch.label).get_indices()
+                        batch_text = batch.text[indices]
+                        batch_label = batch.label[indices]
+                    else:
+                        batch_text = batch.text
+                        batch_label = batch.label
                 else:
-                    scores, rnn_outs = self.model(batch_text, lengths=batch_lengths)
-            else:
-                if 'ignore_lengths' in self.config and self.config['ignore_lengths']:
-                    scores = self.model(batch_text)
+                    if 'resample' in self.config and self.config['resample']:
+                        indices = ImbalancedDatasetSampler(batch.text[0], batch.label).get_indices()
+                        batch_text = batch.text[0][indices]
+                        batch_lengths = batch.text[1][indices]
+                        batch_label = batch.label
+                    else:
+                        batch_text = batch.text[0]
+                        batch_lengths = batch.text[1]
+                        batch_label = batch.label
+
+                if hasattr(self.model, 'tar') and self.model.tar:
+                    if 'ignore_lengths' in self.config and self.config['ignore_lengths']:
+                        logits, rnn_outs = torch.sigmoid(self.model(batch_text)).squeeze(dim=1)
```

Comment:
Is this okay? `rnn_outs` from `torch.sigmoid`?

Reply:
Yeah, this has been a part of the codebase for a while now.